File: 2015/day03_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Position = tuple[int, int]


def solution(content: str) -> Any:
    current: Position = (0, 0)
    visited: list[Position] = [current]
    for direction in content:
        current = move(current, direction)
        visited.append(current)
    return len(
        [True for index, item in enumerate(visited) if item not in visited[:index]]
    )


def move(current: Position, direction: str) -> Position:
    if direction == "^":
        return (current[0], current[1] + 1)
    if direction == "v":
        return (current[0], current[1] - 1)
    if direction == "<":
        return (current[0] - 1, current[1])
    if direction == ">":
        return (current[0] + 1, current[1])
    raise ValueError(f"Invalid direction: {direction}")


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2015/day03-1.hs
import System.IO
import Data.Set (fromList)

type Position = (Int, Int)

solution :: String -> Int
solution = length . fromList . helper [(0,0)]
    where
        helper v [] = v
        helper v (x:xs) = helper ((move (head v) x):v) xs

move :: Position -> Char -> Position
move (x,y) '^' = (x, y+1)
move (x,y) 'v' = (x, y-1)
move (x,y) '<' = (x-1, y)
move (x,y) '>' = (x+1, y)

main :: IO ()
main = do
    handle <- openFile "day03.dat" ReadMode
    content <- hGetContents handle
    print . solution $ content